### ISSUE TITLE
Cricket runs predictor backend code

### ISSUE THREAD

**Issue by sanmarg-debug** [sanmarg-debug/College-projects#2]
import streamlit as st
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

from xgboost import XGBRegressor
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

# Set page config
st.set_page_config(page_title="Cricket Runs Predictor", layout="centered")

# Title
st.title("🏏 Predict Player's Season Runs")

# Load and clean data
@st.cache_data
def load_data():
    df = pd.read_csv("/Users/sanmargdas/Downloads/cricket_data.csv")
    df = df.dropna(subset=['Batting_Average', 'Runs_Scored', 'Batting_Strike_Rate'])

    df['Batting_Average'] = pd.to_numeric(df['Batting_Average'], errors='coerce')
    df['Runs_Scored'] = pd.to_numeric(df['Runs_Scored'], errors='coerce')
    df['Batting_Strike_Rate'] = pd.to_numeric(df['Batting_Strike_Rate'], errors='coerce')
    df = df[df['Batting_Average'] > 0]
    df['Dismissals'] = (df['Runs_Scored'] / df['Batting_Average']).round().astype(int)
    return df

df = load_data()

# Features and target
features = ['Batting_Average', 'Dismissals', 'Batting_Strike_Rate']
X = df[features]
y = df['Runs_Scored']

# Train XGBoost model
model = XGBRegressor(n_estimators=100, learning_rate=0.1, max_depth=4, random_state=42)
model.fit(X, y)

# Sidebar input
st.sidebar.header("🔢 Input Player Stats")
player_name = st.sidebar.text_input("Player Name", "Sample Player")
bat_avg = st.sidebar.number_input("Batting Average", min_value=0.0, value=40.0)
dismissals = st.sidebar.number_input("Dismissals", min_value=0, value=10)
strike_rate = st.sidebar.number_input("Strike Rate", min_value=0.0, value=130.0)

# Prediction
input_df = pd.DataFrame([{
    'Batting_Average': bat_avg,
    'Dismissals': dismissals,
    'Batting_Strike_Rate': strike_rate
}])
predicted_runs = model.predict(input_df)[0]

st.markdown(f"### 🔮 Predicted Season Runs for `{player_name}`: `{predicted_runs:.0f}`")

# Model evaluation
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
y_pred = model.predict(X_test)

mae = mean_absolute_error(y_test, y_pred)
mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

st.markdown("#### 🧪 Model Evaluation Metrics")
st.write(f"**Mean Absolute Error (MAE):** {mae:.2f}")
st.write(f"**Mean Squared Error (MSE):** {mse:.2f}")
st.write(f"**R-squared (R² Score):** {r2:.2f}")

# Visualizations
st.markdown("---")
st.subheader(f"📈 How does `{player_name}` compare?")

# Distribution of Batting Average
fig4, ax4 = plt.subplots()
sns.histplot(df['Batting_Average'], kde=True, ax=ax4, color="skyblue")
ax4.axvline(bat_avg, color='red', linestyle='--', label='Player Input')
ax4.set_title("Distribution of Batting Averages")
ax4.legend()
st.pyplot(fig4)

# Distribution of Strike Rate
fig5, ax5 = plt.subplots()
sns.histplot(df['Batting_Strike_Rate'], kde=True, ax=ax5, color="lightgreen")
ax5.axvline(strike_rate, color='red', linestyle='--', label='Player Input')
ax5.set_title("Distribution of Strike Rates")
ax5.legend()
st.pyplot(fig5)

# Distribution of Dismissals
fig6, ax6 = plt.subplots()
sns.histplot(df['Dismissals'], bins=20, kde=False, ax=ax6, color="lightcoral")
ax6.axvline(dismissals, color='red', linestyle='--', label='Player Input')
ax6.set_title("Distribution of Dismissals")
ax6.legend()
st.pyplot(fig6)

# Player history if available
if 'Player_Name' in df.columns:
    player_data = df[df['Player_Name'].str.lower() == player_name.lower()]
    if not player_data.empty:
        st.subheader(f"📅 {player_name}'s Past Seasons")
        player_input = player_data[features]
        player_data['Predicted_Runs'] = model.predict(player_input)

        fig7, ax7 = plt.subplots()
        sns.lineplot(data=player_data, x='Year', y='Runs_Scored', label='Actual Runs', marker="o", ax=ax7)
        sns.lineplot(data=player_data, x='Year', y='Predicted_Runs', label='Predicted Runs', marker="o", ax=ax7)
        ax7.set_title(f"{player_name} - Actual vs Predicted Runs by Year")
        st.pyplot(fig7)
    else:
        st.info("ℹ️ No historical data available for this player.")

# Correlation heatmap
st.markdown("#### 🔗 Feature Correlation")
fig1, ax1 = plt.subplots()
sns.heatmap(df[features + ['Runs_Scored']].corr(), annot=True, cmap='Blues', ax=ax1)
st.pyplot(fig1)

# Actual vs predicted
st.markdown("#### 🎯 Actual vs Predicted Runs")
fig2, ax2 = plt.subplots()
sns.scatterplot(x=y_test, y=y_pred, ax=ax2)
ax2.set_xlabel("Actual Runs Scored")
ax2.set_ylabel("Predicted Runs")
ax2.plot([y.min(), y.max()], [y.min(), y.max()], 'r--')
st.pyplot(fig2)

# Feature importance
st.markdown("#### 📌 Feature Importance")
importances = model.feature_importances_
fig3, ax3 = plt.subplots()
sns.barplot(x=importances, y=features, ax=ax3)
st.pyplot(fig3)
